zilliqa/tests/it/main.rs

```diff
@@ -278,6 +277,110 @@ impl Network {
         index
     }
 
+    pub fn restart(&mut self) {
+        // We copy the data dirs from the original network, and re-use the same private keys.
+
+        // Note: the tempdir object has to be held in the vector or the OS
+        // will delete it when it goes out of scope.
+        let mut options = CopyOptions::new();
+        options.copy_inside = true;
+
+        // Collect the keys from the validators
+        let keys = self.nodes.iter().map(|n| n.secret_key).collect::<Vec<_>>();
+
+        let validator = (
+            keys[0].node_public_key(),
+            keys[0].to_libp2p_keypair().public().to_peer_id(),
+        );
+        let genesis_committee = vec![validator];
+
+        let (nodes, mut receivers): (Vec<_>, Vec<_>) = keys
+            .into_iter()
+            .enumerate()
+            .map(|(i, key)| {
+                // Copy the persistence over
+                let new_data_dir = tempfile::tempdir().unwrap();
+
+                info!("Copying data dir over");
+
+                if let Ok(mut entry) = fs::read_dir(self.nodes[i].dir.as_ref().unwrap().path()) {
+                    let entry = entry.next().unwrap().unwrap();
+                    info!("Copying {:?} to {:?}", entry, new_data_dir);
+
+                    copy(entry.path(), new_data_dir.path(), &options).unwrap();
+                } else {
+                    warn!("Failed to copy data dir over");
+                }
+
+                node(
+                    genesis_committee.clone(),
+                    None,
+                    key,
+                    i,
+                    Some(new_data_dir),
+                    self.genesis_address,
+                )
+                .unwrap()
+            })
+            .unzip();
+
+        for node in &nodes {
+            trace!(
+                "Node {}: {} (dir: {})",
+                node.index,
+                node.peer_id,
+                node.dir.as_ref().unwrap().path().to_string_lossy(),
+            );
+        }
+
+        let (resend_message, receive_resend_message) =
+            mpsc::unbounded_channel::<(PeerId, Option<PeerId>, Message)>();
+        let receive_resend_message = UnboundedReceiverStream::new(receive_resend_message).boxed();
+        receivers.push(receive_resend_message);
+
+        self.nodes = nodes;
+        self.receivers = receivers;
+        self.resend_message = resend_message;
+
+        // Now trigger a timeout in all of the nodes until we see network activity again
+        // this could of course spin forever, but the test itself should time out.
+        loop {
+            for node in &self.nodes {
+                if node.inner.lock().unwrap().handle_timeout() {
+                    return;
+                }
+                zilliqa::time::advance(Duration::from_millis(1));
```

Comment:
This uses the normal timeout logic that defaults to 5 seconds, right? If so, maybe advancing more than 1ms at a time would still work fine without compromising the test, and reduce the busy spinloop a bit (instead of requiring 5000 iterations every time)?

Reply:
Yes, that's right. I was worried that if you increase the time step too much you could get into weird situations, but that's probably unwarranted. I will bump it to 500ms, seems like a good change